docs: added workflow hooks docs + changed workflow response (#8364)
medusajs/medusa: www/apps/book/app/advanced-development/workflows/add-workflow-hook/page.mdx

export const metadata = {
  title: `${pageNumber} Workflow Hooks`,
}

# {metadata.title}

In this chapter, you'll learn what a workflow hook is, how to add it to your workflow, and how to add workflow hook handlers.

## What is a Workflow Hook?

> A workflow hook is a point in a workflow where you can inject a step function, called a hook handler.

Reviewer: ```suggestion
A workflow hook is a point in a workflow where you can inject custom functionality.
```

> Hook handlers receive an input from the workflow to perform an action using them. They are executed during the workflow's execution when it reaches the hook.

Reviewer: ```suggestion
```

<Note title="Use workflow hooks when" type="success">

- Your workflow is reusable in other applications and you allow performing an external action at some point in your workflow.

</Note>

<Note title="Don't use workflow hooks if" type="error">

- Your workflow isn't reusable by other application. Use a step that performs what a hook handler would instead.

</Note>

---

## How to Add a Hook in a Workflow?

> To add a hook in your workflow, use the `createHook` function imported from `@medusajs/workflows-sdk`.

Reviewer: ```suggestion
To expose a hook in your workflow, use the `createHook` function imported from `@medusajs/workflows-sdk`.
```

  ["13", "createHook", "Add a hook to the workflow."],
  ["14", `"productCreated"`, "The hook's name."],
  ["15", "productId", "The data to pass to the hook handler."],
  ["19", "hooks", "Return the list of hooks in the workflow."]
]

```ts title="src/workflows/my-workflow/index.ts" highlights={hookHighlights}
import {
  createStep,
  createHook,
  createWorkflow,
  WorkflowResponse,
} from "@medusajs/workflows-sdk"
import { createProductStep } from "./steps/create-product"

export const myWorkflow = createWorkflow(
  "my-workflow", 
  function (input) {
    const product = createProductStep(input)
    const productCreatedHook = createHook(
      "productCreated", 
      { productId: product.id }
    )

    return new WorkflowResponse(product, {
      hooks: [productCreatedHook],
    })
  }
)
```

The `createHook` function accepts two parameters:

1. The first is a string indicating the hook's name. This is used to add a hook handler later.
2. The second is the input to pass to the hook handler.

The workflow must also return as a second parameter of `WorkflowResponse` an object having a `hooks` property. Its value is an array of the workflow's hooks.

---

> ## How to Add a Hook Handler?

Reviewer: ```suggestion
## How to consume a Hook?
```

> Hook handlers are created in the `src/workflows/hooks` directory. They use the workflow's `hooks` property to add a hook handler.

Reviewer: ```suggestion
Hook handlers are implemented by registering it on a given workflow in the `src/workflows/hooks` directory of your project. If a workflow exposes hooks, you should see the available ones on the `hooks` property of the workflow.
```
